### PR TITLE
MCOL-2221: Support UTF-8 characters for reverse() sql function

### PR DESCRIPTION
**Pull request by sq2huang**
Change char buffer to wchar_t buffer to support UTF8 characters.

### CODE REVIEW

**Comment by drrtuy** [comment]
Greetings. 
Thank you very much for the contribution. Could you post here that you contribute your patch under BSL license?

File: utils/funcexp/func_reverse.cpp [hunk 1]
```diff
@@ -73,13 +74,15 @@ std::string Func_reverse::getStrVal(rowgroup::Row& row,
...
     // which this one derived.
-    int len = str.length();
-    char* pbuf = new char[len + 1];
-    strncpy(pbuf, str.c_str(), len);
-    pbuf[len] = 0;
-    char* end = pbuf + len - 1;
+    wstring wstr = utf8::utf8_to_wstring(str);
+    int wlen = wstr.length();
+    wchar_t* pbuf = new wchar_t[wlen + 1];
+    wcsncpy(pbuf, wstr.c_str(), wlen);
```
Comment (drrtuy) [inline utils/funcexp/func_reverse.cpp]: I think we should avoid this copy operation sending down 4 pointers into reverse(). begin and end pointers for the source string and two similar for the destination string.

**Comment by sq2huang** [comment]
> Greetings.
> Thank you very much for the contribution. Could you post here that you contribute your patch under BSL license?

I contribute this patch under BSL license

File: utils/funcexp/func_reverse.cpp [hunk 2]
```diff
@@ -73,13 +74,14 @@ std::string Func_reverse::getStrVal(rowgroup::Row& row,
...
-    char* end = pbuf + len - 1;
-    reverse(pbuf, end);
-    string rstr = pbuf;
+    wstring wstr = utf8::utf8_to_wstring(str);
+    int wlen = wstr.length();
+    wchar_t* pbuf = new wchar_t[wlen + 1];
+    pbuf[wlen] = 0;
+    wchar_t* end = pbuf + wlen - 1;
+    reverse(wstr.c_str(), wstr.c_str() + wlen - 1, pbuf, end);
+    wstring wrstr = pbuf;
```
Comment (drrtuy) [inline utils/funcexp/func_reverse.cpp:83]: I wonder if we really need this assignment  or we could use wrstr to instantiate rstr?

Comment (sq2huang) [reply utils/funcexp/func_reverse.cpp:83]: Do you mean get rid of pbuf and write the reversed string directly into wrstr?

**Comment by drrtuy** [comment]
Thanks for the contribution again.

**Review by drrtuy — APPROVED** [review]

**Pull request merged**